File: leetcode/23.2.java
// Runtime: 5 ms, faster than 76.88% of Java online submissions for Merge k Sorted Lists.
// Memory Usage: 41.8 MB, less than 33.34% of Java online submissions for Merge k Sorted Lists.
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

class Solution {

    class NodeCmp implements Comparator<ListNode> {
        public int compare(ListNode a, ListNode b) {
            return a.val - b.val;
        }
    }

    public ListNode mergeKLists(ListNode[] lists) {
        ListNode dummy = new ListNode(0);
        ListNode top = null;
        ListNode cur = dummy;

        //  PriorityQueue<ListNode> pq = new PriorityQueue<>(new NodeCmp());
        PriorityQueue<ListNode> pq = new PriorityQueue<>((ListNode a, ListNode b) ->  a.val - b.val);
        for(ListNode i: lists)
            if(i!=null) pq.add(i);

        while(pq.size()!=0) {
            top = pq.remove();
            cur.next = top;
            cur = cur.next;
            if(top.next != null) pq.add(top.next);
        }

        return dummy.next;
    }
}

File: leetcode/23.2.cpp
//Runtime: 20 ms, faster than 98.81% of C++ online submissions for Merge k Sorted Lists.
//Memory Usage: 11.1 MB, less than 70.24% of C++ online submissions for Merge k Sorted Lists.
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

  class node_cmp{
	  public:
      bool operator()(ListNode *a, ListNode *b) {
		  return a->val > b->val;
	  }
  };

  class Solution {
  public:
	  ListNode* mergeKLists(vector<ListNode*>& lists) {
		  ListNode dummy(0);
		  ListNode* cur = &dummy;
		  ListNode* top = nullptr;

          priority_queue<ListNode*, vector<ListNode*>, node_cmp> pq;
		  for (int i = 0; i < lists.size(); i++) {
			  if (lists[i]) pq.push(lists[i]);
		  }

		  while (!pq.empty()) {
			  top = pq.top();
			  pq.pop();
			  cur->next = top;
              cur = cur->next;
			  if (top->next) pq.push(top->next);
		  }

		  return dummy.next;
	  }
  };